File: bluetooth/peripheral.js
#! /usr/bin/env node
var child_process = require('child_process');
var device_id = child_process.execSync('cat /sys/class/net/eth0/address | sed s/://g').toString().replace(/\n$/, '');

process.env['BLENO_DEVICE_NAME'] = 'LAMPI ' + device_id;

var serviceName = 'LampiService';
var bleno = require('bleno');
var mqtt = require('mqtt');
 
var LampiState = require('./lamp/lampi-state');
var LampiService = require('./lamp/lampi-service');
var DeviceInfoService = require('./device-info-service');
var WifiService = require('./wifi/wifi-service');

var lampiState = new LampiState();
var lampiService = new LampiService( lampiState );
var deviceInfoService = new DeviceInfoService( 'CWRU', 'LAMPI', device_id);
var wifiService = new WifiService();

var bt_clientAddress = null;
var bt_lastRssi = 0;

var mqtt_clientId = 'lamp_bt_central';
var mqtt_client_connection_topic = 'lamp/connection/' + mqtt_clientId + '/state';

var mqtt_options = {
    clientId: mqtt_clientId,
}

var mqtt_client = mqtt.connect('mqtt://localhost', mqtt_options);


bleno.on('stateChange', function(state) {
  if (state === 'poweredOn') {
    //
    // We will also advertise the service ID in the advertising packet,
    // so it's easier to find.
    //
    bleno.startAdvertising('LampiService', [lampiService.uuid, deviceInfoService.uuid], function(err)  {
      if (err) {
        console.log(err);
      }
    });
  }
  else {
    bleno.stopAdvertising();
    console.log('not poweredOn');
  }
});


bleno.on('advertisingStart', function(err) {
  if (!err) {
    console.log('advertising...');
    //
    // Once we are advertising, it's time to set up our services,
    // along with our characteristics.
    //
    bleno.setServices([
        lampiService,
        wifiService,
        deviceInfoService,
    ]);
  }
});

function updateRSSI(err, rssi) {
    // if we are still connected
    if (bt_clientAddress) {
        // and large change in RSSI
        if ( Math.abs(rssi - bt_lastRssi) > 2 ) {
            // publish RSSI value to MQTT 
            mqtt_client.publish('lamp/bluetooth', JSON.stringify({
                'client': bt_clientAddress,
                'rssi': rssi,
                }));
        }
        // update our last RSSI value
        bt_lastRssi = rssi;
        // set a timer to update RSSI again in 1 second
        setTimeout( function() {
            bleno.updateRssi( updateRSSI );
            }, 1000);
        }
}


bleno.on('accept', function(clientAddress) {
    console.log('accept: ' + clientAddress);
    bt_clientAddress = clientAddress;    
    bt_lastRssi = 0;
    mqtt_client.publish('lamp/bluetooth', JSON.stringify({
        state: 'connected',
        'client': bt_clientAddress,
        }));

    bleno.updateRssi( updateRSSI );
});

bleno.on('disconnect', function(clientAddress) {
    console.log('disconnect: ' + clientAddress);
    mqtt_client.publish('lamp/bluetooth', JSON.stringify({
        state: 'disconnected',
        'client': bt_clientAddress,
        }));
    bt_clientAddress = null;
    bt_lastRssi = 0;
});

File: bluetooth/peripheral.py
#!/usr/bin/env python3
from services import *
from pybleno import *
import sys
from helper import Helper
import paho.mqtt.client as mqtt
import os
import json
from services.service_importer import ServiceImporter

TOPIC_BLUETOOTH = 'device/bluetooth'

helper = Helper()

os.environ["BLENO_DEVICE_NAME"] = "MOONLAMP - " + Helper.get_device_id()

bleno = Bleno()

importer = ServiceImporter(helper)
services = importer.inst_classes
utility_service = services['UtilityService']
device_info_service = services['DeviceInfoService']


MQTT_CLIENT_ID = 'device_bt_central'
MQTT_VERSION = mqtt.MQTTv311
MQTT_BROKER_HOST = "localhost"
MQTT_BROKER_PORT = 1883
MQTT_BROKER_KEEP_ALIVE_SECS = 60
mqtt_client = mqtt.Client(client_id=MQTT_CLIENT_ID, protocol=MQTT_VERSION)
mqtt_client.connect(MQTT_BROKER_HOST, port=MQTT_BROKER_PORT, keepalive=MQTT_BROKER_KEEP_ALIVE_SECS)

bt_clientAddress = None
bt_lastRssi = 0


def onStateChange(state):
    print("New state: " + state);

    if (state == 'poweredOn'):
        bleno.startAdvertising('MoonLamp', [utility_service.uuid, device_info_service.uuid])
    else:
        bleno.stopAdvertising()


bleno.on('stateChange', onStateChange)


def onAdvertisingStart(error):
    print('on -> advertisingStart: ' + ('error ' + error if error else 'success'))

    if not error:
        def on_setServiceError(error):
            print('setServices: %s' % ('error ' + error if error else 'success'))
        bleno.setServices(list(services.values()), on_setServiceError)


bleno.on('advertisingStart', onAdvertisingStart)


def updateRSSI(rssi):
    global bt_clientAddress, bt_lastRssi
    if bt_clientAddress:
        if abs(rssi - bt_lastRssi) > 2:
            mqtt_client.publish(TOPIC_BLUETOOTH, json.dumps({
                'client': bt_clientAddress,
                'rssi': rssi
            }))

        bt_lastRssi = rssi



def onAccept(clientAddress):
    global bt_clientAddress, bt_lastRssi
    bt_clientAddress = clientAddress
    bt_lastRssi = 0
    mqtt_client.publish(TOPIC_BLUETOOTH, json.dumps({
        'state': 'connected',
        'client': bt_clientAddress
    }))

    bleno.updateRssi(updateRSSI)

bleno.on('accept', onAccept)


def onDisconnect(clientAddress):
    global bt_clientAddress, bt_lastRssi
    mqtt_client.publish(TOPIC_BLUETOOTH, json.dumps({
        'state': 'disconnected',
        'client': bt_clientAddress
    }))
    bt_clientAddress = None
    bt_lastRssi = 0


bleno.on('disconnect', onDisconnect)

bleno.start()

try:
    while True:
        pass
finally:
    print("running finally")
    helper.db.close()

    bleno.stopAdvertising()
    bleno.disconnect()

    print('Bluetooth daemon terminated.')
    sys.exit(1)


